File: bindings/java/src/main/java/io/uds/UdsClient.java
package io.uds;

import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class UdsClient implements AutoCloseable {
    private long ctx;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final List<Consumer<String>> logListeners = new CopyOnWriteArrayList<>();

    public UdsClient() {
        this.ctx = udsInit();
    }

    public UdsClient(String configPath) {
        this.ctx = udsInitWithConfig(configPath);
    }

    public List<DeviceInfo> listDevices() {
        String json = udsListDevices(ctx);
        return DeviceInfo.fromJsonArray(json);
    }

    public DeviceInfo inspectDevice(String deviceId) {
        String json = udsInspectDevice(ctx, deviceId);
        return DeviceInfo.fromJson(json);
    }

    public CompletableFuture<Boolean> flash(String deviceId, String firmwarePath) {
        return CompletableFuture.supplyAsync(() ->
            udsFlash(ctx, deviceId, firmwarePath) == 0, executor);
    }

    public String rpc(String deviceId, String method, String params) {
        return udsRpc(ctx, deviceId, method, params);
    }

    public void subscribeLogs(Consumer<String> listener) {
        logListeners.add(listener);
    }

    public void close() {
        if (ctx != 0) {
            udsDestroy(ctx);
            ctx = 0;
        }
        executor.shutdown();
    }

    private native long udsInit();
    private native long udsInitWithConfig(String configPath);
    private native void udsDestroy(long ctx);
    private native String udsListDevices(long ctx);
    private native String udsInspectDevice(long ctx, String deviceId);
    private native int udsFlash(long ctx, String deviceId, String firmwarePath);
    private native String udsRpc(long ctx, String deviceId, String method, String params);
}

File: bindings/kotlin/src/main/kotlin/io/uds/UdsClient.kt
package io.uds

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.*
import java.util.concurrent.Executors

class UdsClient(private var ctx: Long = 0) : AutoCloseable {
    private val executor = Executors.newSingleThreadExecutor()
    private val dispatcher = executor.asCoroutineDispatcher()

    constructor(configPath: String? = null) : this(if (configPath != null) udsInitWithConfig(configPath) else udsInit())

    fun listDevices(): List<DeviceInfo> {
        val json = udsListDevices(ctx)
        return Gson().fromJson(json, object : TypeToken<List<DeviceInfo>>() {}.type)
    }

    fun inspectDevice(deviceId: String): DeviceInfo {
        val json = udsInspectDevice(ctx, deviceId)
        return Gson().fromJson(json, DeviceInfo::class.java)
    }

    suspend fun flash(deviceId: String, firmwarePath: String): Boolean =
        withContext(dispatcher) { udsFlash(ctx, deviceId, firmwarePath) == 0 }

    fun rpc(deviceId: String, method: String, params: String): String =
        udsRpc(ctx, deviceId, method, params)

    override fun close() {
        if (ctx != 0L) {
            udsDestroy(ctx)
            ctx = 0L
        }
        executor.shutdown()
    }

    private external fun udsInit(): Long
    private external fun udsInitWithConfig(configPath: String): Long
    private external fun udsDestroy(ctx: Long)
    private external fun udsListDevices(ctx: Long): String
    private external fun udsInspectDevice(ctx: Long, deviceId: String): String
    private external fun udsFlash(ctx: Long, deviceId: String, firmwarePath: String): Int
    private external fun udsRpc(ctx: Long, deviceId: String, method: String, params: String): String
}
